IslandCounter.java

```diff
@@ -0,0 +1,39 @@
+package assignment4;
+
+
+public class IslandCounter {
+
+    private boolean[][] grid;
+
+    public IslandCounter(boolean[][] grid) {
+        if (grid == null || grid.length == 0) {
+            throw new IllegalArgumentException("Grid cannot be empty.");
+        }
+        this.grid = grid;
+    }
+
+    public int countIslands() {

```

Comment:
It would be good to have a complexity analysis here.
If n is the number of grid cells, I think this is an O(n) loop, with each union() call inside the O(n) loop also O(n) for a total complexity of O(n^2).

I think you can solve this problem in O(n) using an algorithm called "flood-fill". You might want to have a look at that algorithm (it is really common) or if you think your algorithm is already then O(n) try to prove it to me (or a reader) by adding javadoc.